bookkeeper-server/src/main/java/org/apache/bookkeeper/http/ReadLedgerEntryService.java

```diff
@@ -0,0 +1,97 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+package org.apache.bookkeeper.http;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Maps;
+import com.google.protobuf.ByteString;
+import java.util.Iterator;
+import java.util.Map;
+import org.apache.bookkeeper.client.BookKeeperAdmin;
+import org.apache.bookkeeper.client.LedgerEntry;
+import org.apache.bookkeeper.conf.ClientConfiguration;
+import org.apache.bookkeeper.conf.ServerConfiguration;
+import org.apache.bookkeeper.http.service.HttpService;
+import org.apache.bookkeeper.http.service.HttpServiceRequest;
+import org.apache.bookkeeper.http.service.HttpServiceResponse;
+import org.apache.bookkeeper.util.JsonUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * HttpService that handle Bookkeeper Configuration related http request.
+ */
+public class ReadLedgerEntryService implements HttpService {
+
+    static final Logger LOG = LoggerFactory.getLogger(ReadLedgerEntryService.class);
+
+    protected ServerConfiguration conf;
+
+    public ReadLedgerEntryService(ServerConfiguration conf) {
+        Preconditions.checkNotNull(conf);
+        this.conf = conf;
+    }
+
+    @Override
+    public HttpServiceResponse handle(HttpServiceRequest request) throws Exception {
+        HttpServiceResponse response = new HttpServiceResponse();
+        Map<String, String> params = request.getParams();
+
+        if (HttpServer.Method.GET == request.getMethod() && (params != null) && params.containsKey("ledger_id")) {
+            Long ledgerId = Long.parseLong(params.get("ledger_id"));
+            Long startEntryId = 0L;
+            Long endEntryId = -1L;
+            if (params.containsKey("start_entry_id")) {
+                startEntryId = Long.parseLong(params.get("start_entry_id"));
+            }
+            if (params.containsKey("end_entry_id")) {
+                endEntryId = Long.parseLong(params.get("end_entry_id"));
+            }
+
+            ClientConfiguration clientconf = new ClientConfiguration(conf)
+              .setZkServers(conf.getZkServers());
+
+            // output <entryid: entry_content>
+            Map<String, String> output = Maps.newHashMap();
+            BookKeeperAdmin bka = new BookKeeperAdmin(clientconf);
+            Iterator<LedgerEntry> entries = bka.readEntries(ledgerId, startEntryId, endEntryId).iterator();
+            while (entries.hasNext()) {
+                LedgerEntry entry = entries.next();
+                output.put(Long.valueOf(entry.getEntryId()).toString(),
+                  ByteString.copyFrom(entry.getEntry()).toStringUtf8());
```

Comment:
This has potential to break the json, if the entry itself also contains json. Why not base64 it?

Reply:
@ivankelly good catch, I don't know if ByteString.copyFrom creates a copy and we can also save resources by just using new String(entry.getEntry(), UTF-8)
anyway if the entry contains json it will be double-encoded.

I would prefer base64 because as entries usually contains WAL data they contain binary strings and they are not made of printable characters or even unmappable utf-8 sequences.

+1 for base64

Reply:
Thanks. will change it to base64.